### ISSUE TITLE
How to use ServerSession get_sni_hostname to switch the ServerConfig to generate new ServerSession and ask client to reinitiate handshake?

### ISSUE THREAD

**Issue by AlexZhao** [ctz/rustls#430]
Hi, 

I'm trying to check how it can be done to re-initiate a ServerSession after got the sni_hostname from previous ServerSession.
for example:
 if I try to use default ServerConfig to new a ServerSession as default ServerSession for client to connect, after client do the handshake and from server side can use ServerSession.get_sni_hostname, then want to switch to the correct ServerConfig according to get_sni_hostname result,  but it seems I try to use ServerSession.set_resumption_data(&[empty]) and write_tls back to client, and seems there is no new TLS handshake initiated from client side, is there any example how to do this?

I think this is quite normal case when we met several https site existed and listening to same IP:PORT with TLS enabled, then server side need to find the relevant ServerConfig according to the application level protocol difference here it is HOST: if there is any example of how to achieve this with rustls it will be quite helpful.

Thanks 
Br
Alex

**Comment by Ralith** [comment]
See also discussion in https://github.com/ctz/rustls/issues/89, which has been blocked on @ctz's feedback for some time.

**Comment by AlexZhao** [comment]
Is it possible to modify the pub struct ResolvesServerCertUsingSNI with #[derive(clone)], then it can be modified on the fly when the server is running.
e.g. 
when the listening point try to use Arc::new(ResolvesServerCertUsingSNI),  if this is cloneable then it is possible to just use clone to create new ServerConfig instance which left existed accepted TCP using existed copy of ServerConfig, and new created can update to new ServerConfig, then when all old TCP connection be release, the previous ServerConfig will be released.

**Comment by ctz** [comment]
There is no standard way to ask a client to reinitiate a handshake, and usually doing this is unsafe.

Could you explain precisely what you're trying to achieve? What parts of ServerConfig do you want to vary based on handshake, that cannot already be arranged for with the existing certificate resolution traits? (see https://docs.rs/rustls/0.19.0/rustls/trait.ResolvesServerCert.html )

**Comment by NeoLegends** [comment]
Hey @AlexZhao,

in our case we hack our way around this by parsing the TLS handshake for the SNI instead (buffering the data), using that to resolve the correct ServerConfig and then finally run rustls on the connection by reconstructing the data stream using e. g. https://docs.rs/tokio/0.3.6/tokio/io/trait.AsyncReadExt.html#method.chain.

That works rather well given that ResolvesServerCert doesn't support asynchronity at the moment.

**Comment by AlexZhao** [comment]
> There is no standard way to ask a client to reinitiate a handshake, and usually doing this is unsafe.
> 
> Could you explain precisely what you're trying to achieve? What parts of ServerConfig do you want to vary based on handshake, that cannot already be arranged for with the existing certificate resolution traits? (see https://docs.rs/rustls/0.19.0/rustls/trait.ResolvesServerCert.html )

Hi,  All

Thanks for your helps, and I try to find a way as below for add new cert to resolver and link it to ServerConfig.

        for server_name in server_names {
            if let None = self.certs.get(server_name) {
                self.certs.insert(server_name.clone(), ck.clone());
            } else {
                warn!("Update the certs which already existed for {}", server_name);
            }
        }

        if let Some(tls_config) = self.tls_config.as_mut() {
            let mut certs_resolve = ResolvesServerCertUsingSNI::new();
            for (name, ck) in self.certs.iter() {
                match certs_resolve.add(name.as_str(), ck.clone()) {
                    Ok(()) => (),
                    Err(e) => {
                        warn!("update cert to resolver failed with error {} for server {}", e, name);
                    }
                }
            }
            
            tls_config.cert_resolver = Arc::new(certs_resolve);
            if let Some(tls_config) = self.tls_config.as_ref() {
                self.arc_tls_config = Some(Arc::new(tls_config.clone()));
            }
        }

Seems it works for me to on the fly add/remove certs from ServerConfig, hope it is a correct way for doing this?

**Comment by ctz** [comment]
I think this is a duplicate of #89, but reopen if you disagree.
